### PR TITLE
[RFC] CP-26024: allowed-operations guided testing

### PR DESCRIPTION
**Pull request by edwintorok**
This is a fairly large set of changes, meant for discussion on the general approach, there are some rough edges around using this, but its already able to find some bugs in the clustering integration.

This refactors code from tests/test_common.ml to be usable as a library,
and defines component level tests for XAPI object operations.
It doesn't delete the code from tests/ to not break existing scripts.

It queries various XAPI object's allowed-operations (e.g. Pool, Host, VDI, Cluster_host, etc.)
and attempts to perform operations sequentially and in parallel, discovering
more available operations along the way.

The goal is to explore operations that a user would be able to do from XenCenter
and from the XE CLI: usually if an operation is claimed to be allowed it should
succeed.
If anything fails that is likely a situation and end-user could stumble upon as well.

This makes manual testing and debugging easier by always having a record of what steps you have executed, when, and what you can do next that can be correlated with the logs.

Overall this would give us some confidence that we didn't miss any obvious action a user would perform, and would be the equivalent of the "exhaustiveness test" in OCaml pattern match for the design and integration.
This approach should be reusable for testing other parts of XenServer in the future, not specific to clustering.

We can implement only as much testing as needed for our immediate goals and extend it as we go along.

It works with both vagrant and xenrt provisioned Xen-on-Xen servers, for
example:
```
jbuilder build
jbuilder exec testarossa_test -- init 10.0.0.10 10.0.0.20 --iscsi 10.0.0.30 --iqn iqn.2009-01.xenrt.test:iscsi18f6f89c  --scsiid 1494554000000000034336135616231630000000000000000 --license-server <internal-license-server> --physical <some-xenrt-machine> --username root --password <password> >config.json
jbuilder exec testarossa_test -- list
jbuilder exec testarossa_test -- prepare --verbosity=debug --config config.json
jbuilder exec testarossa_test -- run --verbosity=debug --config config.json 2>&1 | tee -a test.log
```

### CODE REVIEW

File: lib/allowed_ops.ml [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+            | e -> Lwt.fail e)
+    in
+    rpc ctx B.get_all
+    >>= fun all ->
+    debug (fun m -> m "Performing operations serially") ;
+    Lwt_list.iter_s perform_allowed all
+    >>= fun () ->
+    rpc ctx B.get_all
+    >>= fun all ->
+    debug (fun m -> m "Performing operations in parallel") ;
```
Comment (gaborigloi) [inline lib/allowed_ops.ml]: So this relies on the fact that we usually wait in message forwarding for like 20 mins if there is another operation in progress instead of failing?

**Comment by gaborigloi** [comment]
Really nice! One thing that we could discuss: when we return the allowed operations, we do not know the parameters of the call, we compute them based on the current state. However, when we make a XenAPI call, additional checks are performed based on the parameters. So it's possible that an operation is allowed, but we call it with an invalid param, and it fails with an error.
So we should document how we deal with this. But the way it's done here is nice: we just receive the operation that we should perform, so we are free to choose some parameters that are always guaranteed to be valid.

**Comment by edwintorok** [comment]
Yes, in some places I also had to do some additional checks, like whether the host is disabled or not before reboot/shutdown, I'll add some comments.

**Comment by edwintorok** [comment]
Updated with some comments on how to implement `BEHAVIOUR` module (advising not to duplicate checking done in XAPI), and some bugfixes related to vagrant/ansible/opam files.

File: Vagrantfile [hunk 2]
```diff
@@ -133,7 +134,7 @@ Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
   config.vm.provider "xenserver" do |xs|
     xs.use_himn = true
     xs.memory = 1024
-    xs.xs_host = "gandalf.uk.xensource.com"
+    xs.xs_host = "xrtuk-14-12.xenrt.citrite.net"
```
Comment (gaborigloi) [inline Vagrantfile]: Is this change intentional?
